examples/samv/blink/main.cpp

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2021, Tomasz Wasilczyk
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <modm/architecture/interface/clock.hpp>
+#include <modm/platform.hpp>
+
+using namespace std::chrono_literals;
+
+int
+main()
+{
+//    modm::platform::SystemClock::enable();
+//    SysTickTimer::initialize<SystemClock>();
+
+    using Led0 = modm::platform::GpioD17;
+    using Led1 = modm::platform::GpioD16;
+    using Led2 = modm::platform::GpioD15;
+
+    Led0::setOutput(false);
```

Comment:
I've got to the point where I'm able to build and flash my samv70n20b, but no LEDs blink and after re-plugging debugger it stops responding (and I have to use the ERASE pin to bring it back to life). I know I'm not initializing any clocks, but I assumed GPIO should work without that, am I right?

@mcbridejc would you mind giving me clues where to look next?

Reply:
Do you need to enable the GPIO peripheral clock first?

Reply:
Hi Tomek, I took a look. You don't need to change the master clock settings, but you do need to enable the clock for the PIO devices. I ended up doing this in [gpio/sam/enable.cpp.in](https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/sam/enable.cpp.in), but right now it only happens on "G" families, and it only enabled PIO[A,B]. Looks like the V has A,B,C,D,E. Seems like now we need to create create a list of PIO ports from the device data in module.lb.

Reply:
Actually, this already exists in module.lb: bprops["ports"], and enable.cpp.in already uses this list so you are probably good there. Just need to add the "v" family to the template: 

if target["family"] in ["g", "v"]

Reply:
Thanks for the pointer. Indeed I forgot about PIO and adding "v" to the list generates clock enable code for ports A, B, D (on V70 LQFP100 target). However, still no LEDs light up :).

I tried adding (and not adding) clock config (see the latest revision of this commit) using MAIN_CLK (I don't have external crystal), still no luck. Did I do something else wrong? If you don't have any immediate ideas, I will start doing minimal working build with ASF4 and try bisecting code between it and modm. Moreover, I'm not 100% sure if I'm flashing this correctly with OpenOCD, so there's stuff to verify here as well.

Reply:
I see from main.cpp you are using PIOD pins? If so, I'm afraid pin.hpp.in only supports port A and B at the moment. It will have to be expanded.

See: https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/sam/pin.hpp.in#L135

`getPortReg`, `setPortReg`, and `readPortReg` all have clauses for each port. This should really be a loop in the template, like `%% for port in options["enable_ports"]` in `enable.hpp.in`, but you can add port D real quick and see if that gets you blinking. 

The `toggle` function (see [here](https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/sam/pin.hpp.in#L237)) also has a per-port conditional.

Reply:
So, for example, getPortReg: 

```	
template<PortName port>
inline static constexpr volatile uint32_t*
getPortReg(size_t offset)
{
%% for port in options["enable_ports"]
	if constexpr (port == PortName::{{ port | upper }}) {
		return (uint32_t*)((uint8_t*)PIO{{port | upper}} + offset);
	}
%% endfor
}
```

And similarly for the other 3 functions, and the `PortName` enum in `config.hpp.in`.

Reply:
Also, you have my apologies for not having done it this way already !

Reply:
Yay, it worked! Thanks! Clock/delay is waaay off, but that's manageable/fixable since I know the chip is alive.

Are you about to add the loops or I should do it?

Reply:
Should be
```
%% for port in ports
```
where [`ports` comes from the bports dictionary](https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/sam/module.lb#L41), not the option, o that you can use the API even if you do not enable the ports on startup (you can always do it manually).

Reply:
> Are you about to add the loops or I should do it?

You do it please, there should not be separate PRs for this.

Reply:
> Should be
> 
> ```
> %% for port in ports
> ```
> 
> where [`ports` comes from the bports dictionary](https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/sam/module.lb#L41), not the option, o that you can use the API even if you do not enable the ports on startup (you can always do it manually).

One maybe snag here is that for the `toggle` method to work, you have to [setup the PIO_OWER register](https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/sam/enable.cpp.in#L32) ahead of time, and that's currently done only for ports listed in `enable_ports`, and can only be done after the clock is turned on. This seems like it could be a pretty confusing bug, if someone enables the clock later and then expects the GPIOs to work as normal.

Reply:
Then this limitation must be written into the `enable_ports` option description!

Reply:
Okay, in the most recent version it's done via templates. PTAL at the individual commit (commits for clock and blinky are not ready yet). I had to modify the `module.lb` file in a way that I'm not entirely confident about.